fish-rust/src/common.rs

```diff
@@ -1,11 +1,62 @@
-use crate::ffi;
-use crate::wchar::{wstr, WString};
+use crate::expand::{
+    BRACE_BEGIN, BRACE_END, BRACE_SEP, BRACE_SPACE, HOME_DIRECTORY, INTERNAL_SEPARATOR,
+    PROCESS_EXPAND_SELF, PROCESS_EXPAND_SELF_STR, VARIABLE_EXPAND, VARIABLE_EXPAND_SINGLE,
+};
+use crate::ffi::{self, escape_string_style_t, str2wcstring, wcharz_t};
+use crate::future_feature_flags::{feature_test, FeatureFlag};
+use crate::wchar::{wstr, CharPrefixSuffix, WString};
 use crate::wchar_ext::WExt;
-use crate::wchar_ffi::c_str;
-use crate::wchar_ffi::WCharFromFFI;
+use crate::wchar_ffi::{c_str, WCharFromFFI, WCharToFFI};
+use crate::wildcard::{ANY_CHAR, ANY_STRING, ANY_STRING_RECURSIVE};
+use cxx::{let_cxx_string, CxxWString, UniquePtr};
 use std::os::fd::AsRawFd;
 use std::{ffi::c_uint, mem};
 
+// Highest legal ASCII value.
+pub const ASCII_MAX: char = 127 as char;
+
+// Highest legal 16-bit Unicode value.
+pub const UCS2_MAX: char = '\u{FFFF}';
+
+// Highest legal byte value.
+pub const BYTE_MAX: char = 0xFF as char;
+
+// Unicode BOM value.
+pub const UTF8_BOM_WCHAR: char = '\u{FEFF}';
+
+// Use Unicode "non-characters" for internal characters as much as we can. This
+// gives us 32 "characters" for internal use that we can guarantee should not
+// appear in our input stream. See http://www.unicode.org/faq/private_use.html.
+pub const RESERVED_CHAR_BASE: char = '\u{FDD0}';
+pub const RESERVED_CHAR_END: char = '\u{FDF0}';
+// Split the available non-character values into two ranges to ensure there are
+// no conflicts among the places we use these special characters.
+pub const EXPAND_RESERVED_BASE: char = RESERVED_CHAR_BASE;
+pub const EXPAND_RESERVED_END: char = char_offset(EXPAND_RESERVED_BASE, 16);
+pub const WILDCARD_RESERVED_BASE: char = EXPAND_RESERVED_END;
+pub const WILDCARD_RESERVED_END: char = char_offset(WILDCARD_RESERVED_BASE, 16);
+// Make sure the ranges defined above don't exceed the range for non-characters.
+// This is to make sure we didn't do something stupid in subdividing the
+// Unicode range for our needs.
+// #if WILDCARD_RESERVED_END > RESERVED_CHAR_END
+// #error
+// #endif
```

Comment:
The other option is to use a `#[repr(u32)]` enum and do `Xxx = 'x' as u32,` for each entry, then impl `Into<char> for Enum`. Not an ideal solution, but it would catch any cases of accidentally assigning the same value to two different entries and allow (should we need it) exhaustiveness checking.

Reply:
thanks for the feedback, I've added it to the port of all of common.h in https://github.com/fish-shell/fish-shell/pull/9688